sw/device/tests/sram_ctrl_memset_test.c

```diff
@@ -0,0 +1,103 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This test writes some data to SRAM, sets the `INIT` bit in the `ctrl` CSR,
+//! then tries to read back that written data to ensure the SRAM was scrambled.
+//!
+//! The test currently only checks the retention SRAM so that it can be run out
+//! of executable main SRAM on silicon.
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sram_ctrl_regs.h"  // Generated.
+
+// Define some amount of data we should leave in SRAM to check it gets wiped.
+enum {
+  kSramCtrlTestDataSizeWords = 16,
+};
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static const uint32_t kStatusRegMask = kDifSramCtrlStatusBusIntegErr |
+                                       kDifSramCtrlStatusInitErr |
+                                       kDifSramCtrlStatusEscalated;
+
+static const uint32_t kRandomData[kSramCtrlTestDataSizeWords] = {
+    0x6b4abfae, 0x63bdb6e7, 0x87f99b1a, 0xa214dffe, 0xb12291f9, 0xd0cd1abe,
+    0x5c95e716, 0xe887aab1, 0x307f6ef9, 0x6f5c0464, 0x5882279d, 0x44c19574,
+    0x1bd20079, 0xf8250ead, 0x4bf362a4, 0xad41437d};
+
+/// Write `kRandomData` to an SRAM region.
+static void write_data(mmio_region_t sram_region) {
+  for (int i = 0; i < kSramCtrlTestDataSizeWords; ++i) {
+    mmio_region_write32(sram_region, i * (ptrdiff_t)sizeof(uint32_t),
+                        kRandomData[i]);
+  }
+}
+
+/// Check that the contents of an SRAM region match `kRandomData`.
+static bool read_data(mmio_region_t sram_region) {
+  for (int i = 0; i < kSramCtrlTestDataSizeWords; ++i) {
+    uint32_t word =
+        mmio_region_read32(sram_region, i * (ptrdiff_t)sizeof(uint32_t));
+
+    if (word != kRandomData[i]) {
```

Comment:
I would suggest:
```suggestion
    CHECK(word == kRandomData[i], "Data at index %u mismached,  expected: %x, got: %x");
```
This would provide a batter error message in case the test fails.
Then the function should be called `check_data`.

Reply:
Hmm, I think you're right that this needs a better error message with per-byte `CHECK`s.

It's implemented like this (returning a `bool` and being `CHECK`ed in the `test_main`) because after scrambling, we expect this read to fail, and so want the `CHECK` to fail.

What if the function were `void check_data(mmio_region_t, bool matches)` where `matches == true` runs `CHECK` on every byte, whereas `matches == false` ensures the data does not match `kRandomData` and prints an error if it does?

Reply:
Your suggestion would be exactly what I would do. And I think it's done in some other tests around.

Reply:
Thanks, updated